### PR TITLE
Apply sanity caps on outlier values in performance reports

### LINKED ISSUE #17847: Performance metrics reports incorrect times due to single exceptional high value
## Expected Behavior

Performance reports should use relevant time metrics only. Single outlier and exceptions should not be included in the statistics.

## Current Behavior

Single outliers counting like 4 hours instead of 0.01 seconds totally screw up the metrics.

---

page-performance-metrics

Might be one exceptional high value which is already smoothed by 16 normal values.
page-performance-dom_completion

Might be one exceptional high value which is already smoothed by 189 normal values.
page-performance-dom_procession

---

page-performance-metrics2

One exceptional high value which screws up the whole metrics.
page-performance-download_time

---

## Possible Solution

Only take the fastest/lowest 99th or 99.9th percentile of all values into account. Thereby ignoring the highest 1st or 0.1th percentile of values makes the metrics usable again.

Maybe use Google PageSpeed as orientation: 75th or 95th percentile
https://developers.google.com/speed/docs/insights/v5/about

## Your Environment

* Matomo Version: 4.3.0

## Related issues

Might be related to https://github.com/matomo-org/matomo/issues/17646

### PR DESCRIPTION
**Pull request by rr-it**
### Description:

The values used in performance reports of type 'sum/total' and 'average' are capped to reduce the impact of single failed performance measurements.

This fixes #17847.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

**Comment by sgiehl** [comment]
Hi @rr-it,
Thanks for your contribution. I had a look across the code changes and I'm not totally sure what they are meant for.
To confirm I've understood everything correct:

This PR introduces some sort of maximal values for the performance metrics that are used in calculations.
So if there were values tracked higher than the one configured, the calculations will use the configured values instead.

Is that correct?

**Comment by rr-it** [comment]
@sgiehl That's perfectly correct.

Please also see my latest comment on #17847:
https://github.com/matomo-org/matomo/issues/17847#issuecomment-1546899519

**Comment by rr-it** [comment]
Devices, especially iPhones, sometimes send insane high performance values which don't make any sense.
E.g. DOM processing time of 4 hours instead of 250 microseconds.

## Without this PR
+ 10000 values of 250 ms result in an average of `0.250 seconds`
     $`\frac{10000×250 ms}{10000} = 250 ms`$
+ One value of 4 h and 10000 values of 250 ms result in an average of ~ `1.690 seconds`
  + $`4h = 4×60×60×1000 ms = 14400000 ms`$
  + $`\frac{14400000 ms + 10000×250 ms}{10001} = 1690 ms`$

performance-old

## With this PR
`time_dom_processing_cap_duration_ms = 50000`

+ 10000 values of 250 ms result in an average of `0.250 seconds`
     $`\frac{10000×250 ms}{10000} = 250 ms`$
+ One value of 4 h and 10000 values of 250 ms result in an average of ~ `0.255 seconds`
  + $`4h = 4×60×60×1000 ms = 14400000 ms`$
    $`14400000 ms`$ capped to $`50000 ms`$
  + $`\frac{50000 ms + 10000×250 ms}{10001} = 255 ms`$

performance-new

**Comment by rr-it** [comment]
> If you don't want this PR to be closed automatically in 28 days then you need to assign the label 'Do not close'.

@sgiehl Please add label 'Do not close'.

File: config/global.ini.php
Comment (sgiehl) [inline config/global.ini.php]: Our other plugins are using the plugin name as section only, so guess it makes sense to change that to `PagePerformance` only.

**Review by sgiehl — CHANGES_REQUESTED** [review]
Left a comment. Otherwise the code looks fine to me.
But someone from product team should maybe have a look here first. ping @Stan-vw 
I'm not 100% sure regarding the implementation/solution.
With cutting away values if they are too high I see some possible:
* Only removing too high values, but not also too low values might let look the values better than they are
* The default values for the config values might be inaccurate as some pages might expect higher loading times than others.
* The most accurate solution might be to simply throw away the top 1% and bottom 1% of values and only calculate with the rest. This would be a lot more complex to implement and maybe also slower when archiving, though.

If we plan to include that, guess we would need to update our documentation & faqs accordingly.

**Comment by Stan-vw** [comment]
I think we would like this report to not show the outliers (top & bottom 1% sounds like a fine test to begin with, but as you say that could be more implementation work maybe we can pick fixed values from historic reports).
However, I think we should still have a report that shows all values (if we don't have that already). I've attached some potential ways to visualise this, albeit that I didn't add the relevant axes (days vs seconds):
image
image

**Comment by michalkleiner** [comment]
The business logic still needs to be resolved (whether we do fixed values, top and bottom percentage or something else). There's also a merge conflict now which needs to be resolved, plus a change request from Stefan.

I've removed the Needs review tag for now until all the above is addressed at which point it can be marked as Needs review again.

**Comment by rr-it** [comment]
On the business logic - most points were already mentioned:

+ Clear preference to cap by top percentage over fixed values:
  + But: Top percentage is more implementation work.
  + Fixed values already archive the target to make the chart meaningful. (Right now the charts informational value is insignificant.)
  + There is already a fixed value cap. The Maximum Value for Unsigned MEDIUMINT of database: `16777215 ms ` ~ `4 h 39 min 37 s`
+ Top and bottom? No need for bottom value caps as the impact of low values is neglectable.

## Without this PR

### No outliers

+ 10000 values of 250 ms result in an average of `0.250 seconds`
     $`\frac{10000×250 ms}{10000} = 250 ms`$

### Top outlier

+ One value of 4 h and 10000 values of 250 ms result in an average of ~ `1.690 seconds`
  + $`4h = 4×60×60×1000 ms = 14400000 ms`$
  + $`\frac{14400000 ms + 10000×250 ms}{10001} = 1690 ms`$

### Bottom outlier

+ One value of 1 ms and 10000 values of 250 ms result in an average of ~ `0.250 seconds`
  + $`\frac{1 ms + 10000×250 ms}{10001} = 249.975 ms`$

**Comment by michalkleiner** [comment]
@Stan-vw I guess it's your call on whether we should accept this as is or you want any adjustments.

**Comment by Stan-vw** [comment]
I understand the implication of the math and agree that the reports' value/reliability diminishes drastically if outliers are included in the calculation. My challenges with giving a clear conclusive answer are:
1. I am not yet familiar with the report we're talking about, and the different ways users might be leveraging it. I'm wondering what type of performance are people tracking: Website pageloads are in milliseconds, but what if people are tracking items similar to our archiving processes which can take much longer? 
2. If we are setting the cut-off values, they ultimately are arbitrary for users. Can we make the lower/upper values a variable they can set themselves? (either per report or as a whole)
3. One of our pillars is to show all data so I'm hesitant to simply greenlight a change that hides some data, especially when we're not telling users about it. Can we add something like "X1 data points were hidden as they are below the outlier threshold of Y1 milliseconds, X2 data points were hidden as they are above the outlier threshold of Y2 milliseconds"?
4. Given that people might have an edge case where outliers on either side are of interest, can we give users the ability to turn this off in a config?
5. Would it be a decent idea to make this a separate report, so that people can put "Performance report" and "Performance report (excluding outliers)" next to each other?

Thanks in advance for helping me understand the feature and how people are using it more deeply.

**Comment by Stan-vw** [comment]
Thanks @michalkleiner for pointing me to the [area of this PR that answers all my questions](https://github.com/matomo-org/matomo/pull/20718/files#diff-72968590ea965bba4c1ecf90875001330282aeaa9278c468336a96a6d683ab0e)

If I understand correctly, the current PR already allows people to:
1. Choose their own values
2. Turn this feature on/off

Knowing this, I am happy for this to be added as an opt-in feature (from @sgiehl 's message above I interpret it's gone through review and testing).

In the current state (as far as I understand it), I would not ship this as a default state because people aren't aware it's hiding data. I think it could be appropriate as a default state when we can clearly indicate how many datapoints were excluded from the calculation of a metric in some user friendly way (some non-vetted ideas that comes to mind: show "avg load time of 250 ms, excluding 3 outliers above 10000 ms" on hover-over or on a footnote).

**Comment by michalkleiner** [comment]
Thanks @Stan-vw, that's great. I also think making this an opt-in measure addresses all the concerns here and we can always iterate on that in future, while on-premise users will be able to configure this locally as they please.

@sgiehl what is the best way to provide the example config but not have it active, would it still be in the `global.ini.php` file but just commented out? Then the comment about disabling it via providing 0 values can be removed and the code adjusted to treat missing config as this being disabled.

**Comment by Stan-vw** [comment]
This might be interesting to add in the general settings? Have added a quick & dirty mockup of what this could look like. Not sure if that'd work though.
image

File: plugins/PagePerformance/Columns/Base.php
Comment (michalkleiner) [inline plugins/PagePerformance/Columns/Base.php]:
```suggestion
    private function getConfigValueCap()
```

Comment (sgiehl) [reply plugins/PagePerformance/Columns/Base.php]: Does private work when the method is called from inherited classes?

Comment (rr-it) [reply plugins/PagePerformance/Columns/Base.php]: We can go for `private` here as `Base->getConfigValueCap` gets only called by `Base->getSqlCappedValue`.

If we need to call `Base->getConfigValueCap` from functions _declared in_ an inherited class, we could use `protected` instead of `private`.

Also see: https://www.php.net/manual/en/language.oop5.visibility.php

Comment (michalkleiner) [reply plugins/PagePerformance/Columns/Base.php]: Yep, `private` here is fine, checked the usages. The other method needs to stay public as it's called from other classes, too.

**Comment by bx80** [comment]
iPhone outliers could be related to different behaviour with the DOMContentLoaded event on IOS https://developer.apple.com/forums/thread/651215

**Comment by rr-it** [comment]
> iPhone outliers could be related to different behaviour with the DOMContentLoaded event on IOS https://developer.apple.com/forums/thread/651215

All performance metrics are just read from [PerformanceTiming API](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceTiming) e.g. [loadEventEnd](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceTiming/loadEventEnd)

**Comment by bx80** [comment]
>All performance metrics are just read from [PerformanceTiming API](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceTiming) e.g. [loadEventEnd](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceTiming/loadEventEnd)

Thanks @rr-it, I was curious about the cause of the high DOM loading times, but I now see that issue https://github.com/matomo-org/matomo/issues/17646 you linked in the related other ticket gives some idea of how these outlying values can originate :+1:

**Comment by michalkleiner** [comment]
Hi @rr-it,
thanks again for providing the PR, we do appreciate it. At the moment we are thinking whether the configuration should be done via a config or, as @Stan-vw mocked above, through the settings, so that people who don't have access to modifying the config files were able to also adjust the values.
Would that we an option that you'd be willing to look into as well?

We would also need to have the mechanism covered by a test to know that the caps are working correctly for future maintainability, are you familiar with how to provide one?

**Comment by michalkleiner** [comment]
Hi @rr-it, just checking in if you were still keen to progress this PR adding the UI for the values configuration?

**Comment by rr-it** [comment]
Hi @michalkleiner, 

On the UI for configuration:
I don't have the capacity to implement this and I never looked into the code of the Matomo config UI.

On Unit Tests:
Unfortunately I did not get the tests running in my local development environment (Debian Linux).
Next I gave it a try on DDEV/Docker Container - but still I could not easily resolve the dependencies.

**Comment by michalkleiner** [comment]
@Stan-vw it would be good to get this over the line — add the settings section to configure the values through the UI and add a test, it's not too far off being mergable and it's a nice community contribution.

**Comment by Stan-vw** [comment]
@michalkleiner it's been ages since we discussed this so I could be wrong, but I believe we agreed to merge this without front end configuration option for now. I'm actually surprised to see it's still open rather than merged, what part is not mergable yet?

**Comment by rr-it** [comment]
@sgiehl You *merged* branch '5.x-dev' into `issue/17847-performance-report-cap-values`.

If you like I can revert this merge and instead *rebase* my changes onto branch '5.x-dev'. And then do a *force push* of this newly created branch for `issue/17847-performance-report-cap-values`.

Done: I proceeded as described above.

**Comment by rr-it** [comment]
The following code throws WARNING/NOTICE in "Matomo Tests / UI (0-3)":

https://github.com/rr-it/matomo/blob/27e6b666b01eea12acefec6271622ed90716ddc9/plugins/PagePerformance/Columns/Base.php#L29-L34
```php
        try {
            $valueCap = Config::getInstance()->PagePerformance[$this->columnName . '_cap_' . $this->type];
        } catch (Exception $ex) {
            // 0 disables cap
            return 0;
        }
```

The WARNING/NOTICE is like:
```
2024-01-03T15:31:12.7143577Z WARNING PagePerformance
[2024-01-03 15:24:08 UTC] [5529] /home/runner/work/matomo/matomo/matomo/plugins/PagePerformance/Columns/Base.php(30):
Notice - Undefined index: time_network_cap_duration_ms - Matomo 5.0.0 - Please report this message in the Matomo forums: https://forum.matomo.org (please do a search first as it might have been reported already)
#0/plugins/PagePerformance/Columns/Base.php(30),
#1/plugins/PagePerformance/Columns/Base.php(40),
#2/plugins/PagePerformance/Metrics.php(115),
#3/plugins/PagePerformance/PagePerformance.php(151),
[internal function]: Piwik\Plugins\PagePerformance\PagePerformance->addActionMetrics(),
#5/core/EventDispatcher.php(147),
#6/core/Piwik.php(870),
#7/plugins/Actions/Metrics.php(89),
#8/plugins/Actions/RecordBuilders/ActionReports.php(194),
#9/plugins/Actions/RecordBuilders/ActionReports.php(81)
```

How do you correctly call `Config::getInstance()->PagePerformance['this_index_might_be_not_set'];` without getting a WARNING/NOTICE?

**Comment by rr-it** [comment]
On runs of "Matomo Tests / UI (0-3)" the php-log does not become cluttered anymore with PHP `NOTICE`-entries.

The added `?? 0` does the trick to hide `NOTICE: undefined index`. And it also ensures that on missing configuration the default becomes `0`:
`$valueCap = Config::getInstance()->PagePerformance['this_index_might_be_not_set'] ?? 0;`

@sgiehl Now I think this PR is ready to merge.

File: config/global.ini.php
Comment (sgiehl) [inline config/global.ini.php]: I would set all values to 0 by default. Having the lines commented out will hide the from the configuration overview page in Matomo. Providing a default value of 0 will show them there and make that possibility a bit more visible.

```suggestion
; The configuration below provides the possibility to enable capping of values used for generating 'sum/total' and 'average' metrics for page performance reports.
; This allows to reduce the impact of single failed performance measurements.
;
; The recommended caps are at 100-fold an avg/high value. Thereby one wrong value in 10,000 values results in a 1% deviation:
; (1x 100N + 9999x N) / 10000 ~= 101% N
;
; By default capping is disabled. To enable capping overwrite the configs below with a value higher than 0.

; Cap for Network time: avg/high 100ms (recommended value: 10000)
time_network_cap_duration_ms = 0

; Cap for Server time: avg/high 500ms (recommended value: 50000)
time_server_cap_duration_ms = 0

; Cap for Transfer time: avg/high 250ms (recommended value: 25000)
time_transfer_cap_duration_ms = 0

; Cap for DOM processing time: avg/high 500ms (recommended value: 50000)
time_dom_processing_cap_duration_ms = 0

; Cap for DOM completion time: avg/high 1500ms (recommended value: 150000)
time_dom_completion_cap_duration_ms = 0

; Cap for On load time: avg/high 10ms (recommended value: 1000)
time_on_load_cap_duration_ms = 0

```

**Review by sgiehl — CHANGES_REQUESTED** [review]
Sorry for not coming back earlier on this one. I would suggest to update the configuration slightly. Otherwise this imho should look good. We could though consider to add some tests around that, to ensure it works correctly.

File: config/global.ini.php
Comment (rr-it) [reply config/global.ini.php]: > Having the lines commented out will hide the from the configuration overview page in Matomo. Providing a default value of 0 will show them there and make that possibility a bit more visible.

Nice find! I didn't know about that.

**Review by sgiehl — APPROVED** [review]
I've added a simple integration test now to cover the value capping in a test. Imho this should be good to merge now. We might need to update the config ui test after merging though

File: plugins/PagePerformance/tests/Integration/CappedValuesTest.php
Comment (michalkleiner) [inline plugins/PagePerformance/tests/Integration/CappedValuesTest.php:66]: @sgiehl should this be called `testShouldCapOutlinerValuesWhenConfigured()` without the 'not'?

Comment (sgiehl) [reply plugins/PagePerformance/tests/Integration/CappedValuesTest.php:66]: indeed

**Pull request merged**
